### PR TITLE
contrib: add docker-machine-install-bundle.sh

### PR DESCRIPTION
**Pull request by AkihiroSuda**
**\- What I did**
Added `docker-machine-install-bundle.sh`, which installs the bundle to Docker Machine instances.
This script helps developers to test the latest Docker with Swarm-mode enabled.
It is not designed to be used in production.

In future we should support [Infrakit](https://github.com/docker/infrakit) and [Testkit environment](https://github.com/docker/docker-e2e/tree/master/testkit/environment) as well.

**\- How I did it**

Copies `bundles/$VERSION/binary-{client,daemon}/{docker,dockerd,..}` to an instances's `/usr/bin` via `docker-machine ssh`.
It installs the files under `dynbinary-{client,daemon}` instead if found.

**\- How to verify it**

``` console
$ docker-machine create -d some-driver dm01
..
$ docker-machine create -d some-driver dm05
$ make binary
$ contrib/docker-machine-install-bundle.sh install dm01 dm02 dm03 dm04 dm05
Files to be installed:
bundles/1.13.0-dev/binary-daemon/dockerd
bundles/1.13.0-dev/binary-daemon/docker-proxy
bundles/1.13.0-dev/binary-daemon/docker-containerd
bundles/1.13.0-dev/binary-daemon/docker-containerd-ctr
bundles/1.13.0-dev/binary-daemon/docker-containerd-shim
bundles/1.13.0-dev/binary-daemon/docker-init
bundles/1.13.0-dev/binary-daemon/docker-runc
bundles/1.13.0-dev/binary-client/docker
dm01: stopping docker
dm02: stopping docker
dm04: stopping docker
dm03: stopping docker
dm05: stopping docker
dm05: installing docker
dm01: installing docker
dm04: installing docker
dm05: starting docker
dm01: starting docker
dm04: starting docker
dm05: done
dm01: done
dm04: done
dm03: installing docker
dm03: starting docker
dm03: done
dm02: installing docker
dm02: starting docker
dm02: done
$ eval (docker-machine env dm01)
$ docker version
```

**\- Description for the changelog**

contrib: add docker-machine-install-bundle.sh

**\- A picture of a cute animal (not mandatory but encouraged)**

Signed-off-by: Akihiro Suda suda.akihiro@lab.ntt.co.jp

### CODE REVIEW

**Comment by AkihiroSuda** [comment]
cc @justincormack @dperny @aluzzardi

**Comment by thaJeztah** [comment]
ping @tianon PTAL

**Comment by AkihiroSuda** [comment]
cc @nathanleclaire

**Comment by nathanleclaire** [comment]
Ah, very cool.  I am not maintainer on this repository, so I can't bless code for merge, but the script seems sound for the intended use case.

I have always thought a `docker-machine daemon` command to manage the daemon in a cross-init-system way would be nice, so that e.g., `docker-machine daemon restart` could replace `docker-machine ssh sudo systemctl restart docker` in this script.

File: contrib/docker-machine-install-bundle.sh [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+    # TODO: compression (should not be default)
+    tar ch --transform 's/.*\///' $files | docker-machine ssh $m sudo tar Cx $PREFIX/bin
+    echo "$m: starting docker"
+    start_docker $m
+    echo "$m: done"
+}
+
+check_prereq(){
+    command -v docker-machine > /dev/null || errexit "docker-machine not installed"
+    ( tar --version | grep GNU > /dev/null ) || errexit "GNU tar not installed"
```
Comment (nathanleclaire) [inline contrib/docker-machine-install-bundle.sh]: Good to see folks not just blindly assuming GNU tar :)

Comment (AkihiroSuda) [inline contrib/docker-machine-install-bundle.sh]: :smile:

**Comment by AkihiroSuda** [comment]
updated PR

 * Now it can fallbacks to `service` if `systemctl` not found
 * Now it automatically detects the prefix (e.g. `/usr`)

The PR is not likely to included in 1.13 (freezed).
However, I updated PR now because I think this script is useful for developers and testers :smile:

File: contrib/docker-machine-install-bundle.sh [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+	echo $BUNDLE/binary-client/docker
+    fi
+}
+
+control_docker(){
+    m=$1; op=$2
+    # NOTE: `docker-machine ssh $m sh -c "foo bar"` does not work
+    #       (but `docker-machine ssh $m sh -c "foo\ bar"` works)
+    #       Anyway we avoid using `sh -c` here for avoiding confusion
+    script="if hash systemctl 2> /dev/null; then systemctl $op docker; else service docker $op; fi"
```
Comment (tianon) [inline contrib/docker-machine-install-bundle.sh]: Can we add a fallback to `/etc/init.d/docker` so this can work on boot2docker too? :pray: :innocent:

Comment (tianon) [inline contrib/docker-machine-install-bundle.sh]: Indeed, although there is a symlink included at `/etc/init.d/docker` for better compatibility/discoverability: https://github.com/boot2docker/boot2docker/blob/bc7888a03a6cb0480bfdf31a7b615e516d0afe52/rootfs/make_iso.sh#L17 :smile: :+1:

Comment (AkihiroSuda) [inline contrib/docker-machine-install-bundle.sh]: oh, thank you for notifying that :+1: 
Updated PR

**Review by tianon — APPROVED** [review]
:+1:

**Review by thaJeztah — APPROVED** [review]
LGTM!

**Pull request merged**